File: LeetCode/101-200/189. Rotate Array/src/Solution.java
import java.util.ArrayDeque;
import java.util.Queue;

class Solution {
    public void rotate(int[] nums, int k) {
        Queue<Integer> q = new ArrayDeque<>();
        k = nums.length - (k % nums.length);
        for (int i = 0; i < nums.length; i++) {
            q.add(nums[i]);
            if (i + k < nums.length) {
                nums[i] = nums[i + k];
            } else {
                nums[i] = q.poll();
            }
        }
    }
}

File: LeetCode/101-200/189. Rotate Array/Solution.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k = n - (k % n);
        if (k == 0) {
            return;
        }
        for (int i = 0; i < k / 2; ++i) {
            swap(nums[i], nums[k - i - 1]);
        }
        for (int i = k; i < (n + k) / 2; ++i) {
            swap(nums[i], nums[n - i + k - 1]);
        }
        reverse(nums.begin(), nums.end());
    }
};